src/cow-react/modules/wallet/web3-react/connectors/Injected/index.tsx

```diff
@@ -58,17 +61,24 @@ export class InjectedWallet extends Connector {
         ? desiredChainIdOrChainParameters
         : desiredChainIdOrChainParameters?.chainId
 
-    if (this.connected && desiredChainId && desiredChainId === this.parseChainId(this.provider.chainId)) {
-      const desiredChainIdHex = `0x${desiredChainId.toString(16)}`
+    return Promise.all([this.provider.request({ method: 'eth_chainId' }) as Promise<string>, this.getAccounts()]).then(
+      ([chainId, accounts]) => {
+        const receivedChainId = this.parseChainId(chainId)
+
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        if (!desiredChainId || desiredChainId === receivedChainId) {
+          return this.actions.update({ chainId: receivedChainId, accounts })
+        }
 
-      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-      return this.provider
-        .request<void>({
+        const desiredChainIdHex = `0x${desiredChainId.toString(16)}`
+
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        return this.provider!.request<void>({
           method: 'wallet_switchEthereumChain',
           params: [{ chainId: desiredChainIdHex }],
-        })
-        .catch(async (error: ProviderRpcError) => {
+        }).catch(async (error: ProviderRpcError) => {
           console.log('err!', error)
+          console.log('debug err', error)
```

Comment:
Can you make some nicer error messages?

Reply:
also, they are duplicated and it should be console.error

On top of this, should we print if we also throw in some cases?

Reply:
My mistake, this was left from me debugging, I will do what web3-react is doing [here](https://github.com/Uniswap/web3-react/blob/de97c00c378b7909dfbd8a06558ed12e1f796caa/packages/metamask/src/index.ts#L160)